src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -204,6 +205,51 @@ protected Solution SetCurrentSolution(Solution solution)
             }
         }
 
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/> and updates <see cref="CurrentSolution"/> to the new value.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="oldSolution">Previous instance of the solution.</param>
+        /// <param name="newSolution">Transformed instance of the solution with updated <see cref="Solution.WorkspaceVersion"/>.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        protected bool SetCurrentSolution(Func<Solution, Solution> transformation, out Solution oldSolution, out Solution newSolution)
+        {
+            if (transformation == null)
+            {
+                throw new ArgumentNullException(nameof(transformation));
+            }
+
+            var currentSolution = Volatile.Read(ref _latestSolution);
+
+            while (true)
+            {
+                var transformedSolution = transformation(currentSolution);
+                if (transformedSolution == currentSolution)
+                {
+                    oldSolution = newSolution = currentSolution;
+                    return false;
+                }
+
+                newSolution = transformedSolution.WithNewWorkspace(this, currentSolution.WorkspaceVersion + 1);
+                var replacedSolution = Interlocked.CompareExchange(ref _latestSolution, newSolution, currentSolution);
+                if (replacedSolution == currentSolution)
+                {
+                    oldSolution = replacedSolution;
+                    return true;
+                }
+
+                currentSolution = replacedSolution;
+            }
+        }
+
+        internal bool SetCurrentSolutionWithLock(Func<Solution, Solution> transformation, out Solution oldSolution, out Solution newSolution)
```

Comment:
if we have SetCurrentSolutionWithLock do we need the lock-free version above?

Reply:
This mirrors `SetCurrentSolution` and is internal and thus to be called only with transformation we understand will not cause deadlock.
Ideally we would remove this. But currently all solution updating methods take the lock, so we need to take it as well.

Reply:
https://github.com/dotnet/roslyn/issues/42237

Reply:
Yeah, I'm not getting this either. There's no real value in any derived class with it's own locking to call this (say, VisualStudioWorkspace), and what benefit does it bring for HandleProjectChange? If nothing else, I'd appreciate a long commit comment explaining the motivation here.

Reply:
I'm not saying this is _wrong_, but if this is where commit messages and PR messages are really important to clarify some intent here.

Reply:
OK, let me clarify and I'll also add more comments - I guess I just thought about it too much already so it looks more obvious to me than it actually is :).

We have `SetCurrentSolution` that atomically updates the Workspace to a given solution while making also sure that the version of the solution will be updated. This is a protected method that derived classes may use to update current solution.

We also have many methods on Workspace that read CurrentSolution make an update and then call `SetCurrentSolution`. To make this combined operation atomic they use `_serializationLock`. `HandleProjectChange` is one of the places this is done used by `OnXxxChanged` methods. The problem with the current pattern is that the lock is held while calling to external code (e.g. raising an event, or calling various virtual  protected methods on Workspace). 

I am not trying to address all of these in entirety since that's a lot of work. Rather, in this PR I'm primarily trying to simplify callers of `ApplyChangeToWorkspace`.
I'm gonna follow up on the above issues in separate PRs later on - moving the event outside of the lock will address https://github.com/dotnet/roslyn/issues/42229. 

All of these will use the new `SetCurrentSolution` overload that takes a solution transformation function (like a silver bullet :)). This new overload behaves exactly like the existing overload but with the extra transformation done on the solution before it is set to the Workspace. This makes it possible to transform the current solution to a new one atomically without taking the lock. Until however all the methods on workspace that currently use the `_serializationLock` switch to this new overload we still need to take the lock when updating the CurrentSolution to synchronize between the old style code and the new helper. That's why I added `SetCurrentSolutionWithLock`.

After this is merged we can migrate the old pattern to use `SetCurrentSolutionWithLock`. Once `SetCurrentSolutionWithLock` is the only place that uses the `_serializationLock` we can remove the lock. Tracked by https://github.com/dotnet/roslyn/issues/42237. I tried to do that but looks like a quite a bit of work because of all the virtual protected methods that can't be changed without potentially breaking something. So, while it would be nice to do this completely it is not the highest priority right now. Why to add the new overload then? It makes some things simpler and we can use it for newly added solution operations (I need to add a few) and avoid adding more virtual protected methods that use the broken pattern and might make it harder to do this 100% later.

BTW, I made the new overload `protected`, so that it can be used by whoever sub-classes Workspace, but I can also keep it `internal` for now and only make it protected when someone actually asks.

Makes sense?

Reply:
PR addressing #42229: https://github.com/dotnet/roslyn/pull/42310

Reply:
Turns out moving the event triggering out of under the lock is not enough and I'll need to move some OnXXXChanges methods to use SetCurrentSolution with transformation helper to avoid the deadlock.

Reply:
> BTW, I made the new overload protected, so that it can be used by whoever sub-classes Workspace, but I can also keep it internal for now and only make it protected when someone actually asks.

I'd absolutely say do that. Some of this mess we're in is because we were fairly lazy about making things protected without really thinking through the impacts. I'd hate for us to make some protected methods now that we'll be transitioning away from and then have to support them again in some way.

Reply:
Will do.